pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,273 @@
+	// generate the cache key based on query, tenant and start time.
+	cacheKey := fmt.Sprintf("log:%s:%s:%d", tenant.JoinTenantIDs(tenantIDs), req.GetQuery(), alignedStart.UnixNano()/(interval.Nanoseconds()))
+
+	_, buff, _, err := l.cache.Fetch(ctx, []string{cache.HashKey(cacheKey)})
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error fetching cache", "err", err, "cacheKey", cacheKey)
+		return l.next.Do(ctx, req)
+	}
+	// we expect only one key to be found or missing.
+	if len(buff) > 1 {
+		level.Warn(l.logger).Log("msg", "unexpected length of cache return values", "buff", len(buff))
+		return l.next.Do(ctx, req)
+	}
+
+	if len(buff) == 0 {
+		// cache miss
+		return l.handleMiss(ctx, cacheKey, lokiReq)
+	}
+
+	// cache hit
+	var cachedResquest LokiRequest
+	err = proto.Unmarshal(buff[0], &cachedResquest)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error unmarshalling request from cache", "err", err)
+		return l.next.Do(ctx, req)
+	}
+	return l.handleHit(ctx, cacheKey, &cachedResquest, lokiReq)
+}
+
+func (l *logResultCache) handleMiss(ctx context.Context, cacheKey string, req *LokiRequest) (queryrangebase.Response, error) {
+	l.metrics.CacheMiss.Inc()
+	level.Debug(l.logger).Log("msg", "cache miss", "key", cacheKey)
+	resp, err := l.next.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	lokiRes, ok := resp.(*LokiResponse)
+	if !ok {
+		return nil, fmt.Errorf("unexpected response type %T", resp)
+	}
+	// we have data we don't want to cache.
+	if !isEmpty(lokiRes) {
+		return resp, nil
+	}
+	data, err := proto.Marshal(req)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error marshalling request", "err", err)
+		return resp, nil
+	}
+	// cache the result
+	err = l.cache.Store(ctx, []string{cache.HashKey(cacheKey)}, [][]byte{data})
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error storing cache", "err", err)
+	}
+	return resp, nil
+}
+
+func (l *logResultCache) handleHit(ctx context.Context, cacheKey string, cachedResquest *LokiRequest, lokiReq *LokiRequest) (queryrangebase.Response, error) {
+	l.metrics.CacheHit.Inc()
+	// we start with an empty response
+	result := emptyResponse(cachedResquest)
+	// if the request is the same and cover the whole time range.
+	// we can just return the cached result.
+	if lokiReq.GetStartTs().After(cachedResquest.GetStartTs()) || lokiReq.GetStartTs().Equal(cachedResquest.GetStartTs()) &&
+		lokiReq.GetEndTs().Before(cachedResquest.GetEndTs()) || lokiReq.GetEndTs().Equal(cachedResquest.GetEndTs()) {
+		return result, nil
+	}
+	// we could be missing data at the start and the end.
+	// so we're going to fetch what is missing.
+	var (
+		startRequest, endRequest *LokiRequest
+		startResp, endResp       *LokiResponse
+		updateCache              bool
+	)
+	g, ctx := errgroup.WithContext(ctx)
+
+	if lokiReq.GetStartTs().Before(cachedResquest.GetStartTs()) {
+		g.Go(func() error {
+			startRequest = lokiReq.WithStartEndTime(lokiReq.GetStartTs(), cachedResquest.GetStartTs())
```

Comment:
it looks like you have `cachedResquest` in a few places :)

Reply:
Yes it's intentional I'm fetching what is missing. From the req to what is in the cache.

Reply:
I also thought this is a typo and still have a feeling it might be. Maybe it needs to be clearer what the extra `s` stands for?

```suggestion
			startRequest = lokiReq.WithStartEndTime(lokiReq.GetStartTs(), cachedRequest.GetStartTs())
```

Reply:
It was a typo.